olm/catalogd/v1alpha1/clustercatalog_types.go

```diff
@@ -0,0 +1,223 @@
+	// +optional
+	URLs *ClusterCatalogURLs `json:"urls,omitempty"`
+	// lastUnpacked represents the time when the
+	// ClusterCatalog object was last unpacked successfully.
+	// +optional
+	LastUnpacked metav1.Time `json:"lastUnpacked,omitempty"`
+}
+
+// ClusterCatalogURLs contains the URLs that can be used to access the catalog.
+type ClusterCatalogURLs struct {
+	// base is a required cluster-internal URL which provides API access for this ClusterCatalog.
+	// A suffix API access path can be added to retrieve catalog data for the ClusterCatalog.
+	// Currently, a 'v1' API access provides complete FBC retrival via the path "/api/v1/all", with the general form `{base}/api/v1/all`.
+	// +kubebuilder:validation:Required
+	Base string `json:"base"`
+}
+
+// CatalogSource is a discriminated union of possible sources for a Catalog.
+// CatalogSource contains the sourcing information for a Catalog
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type == 'Image' && has(self.image)",message="source type 'Image' requires image field"
+type CatalogSource struct {
+	// type is a required reference to the type of source the catalog is sourced from.
+	//
+	// Allowed values are ["Image"]
+	//
+	// When this field is set to "Image", the ClusterCatalog content will be sourced from an OCI image.
+	// When using an image source, the image field must be set and must be the only field defined for this type.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:="Image"
+	// +kubebuilder:validation:Required
+	Type SourceType `json:"type"`
+	// image is used to configure how catalog contents are sourced from an OCI image. This field must be set when type is set to "Image" and must be the only field defined for this type.
+	// +optional
+	Image *ImageSource `json:"image,omitempty"`
+}
+
+// ResolvedCatalogSource is a discriminated union of resolution information for a Catalog.
+// ResolvedCatalogSource contains the information about a sourced Catalog
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type == 'Image' && has(self.image)",message="source type 'Image' requires image field"
+type ResolvedCatalogSource struct {
+	// type is a reference to the type of source the catalog is sourced from.
+	//
+	// It will be set to one of the following values: ["Image"].
+	//
+	// When this field is set to "Image", information about the resolved image source will be set in the 'image' field.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:="Image"
+	// +kubebuilder:validation:Required
+	Type SourceType `json:"type"`
+	// image is a field containing resolution information for a catalog sourced from an image.
+	Image *ResolvedImageSource `json:"image"`
+}
+
+// ResolvedImageSource provides information about the resolved source of a Catalog sourced from an image.
+type ResolvedImageSource struct {
+	// ref contains the resolved sha256 image ref containing Catalog contents.
+	Ref string `json:"ref"`
```

Comment:
In theory, this ref format would need to grow to accommodate other enums that show up in the `spec.source` discriminated union.

Would you suggest validating as an image reference for now, and expanding the validation if/when new enums are added? Or should we leave it open to avoid changing the validation (as @deads2k says, all changes are breaking changes)?

Reply:
Nevermind, I see that we have a matching discriminated union in the status. So, yes, this is always planned to be a digest-based image ref.

But still similar question as above: Do we want to specifically enforce a format for the digest-component of the ref? If OCI ever grows support for new digest algorithms (e.g. `sha256` is found to be vulnerable to collision attacks), we'll need to expand our pattern to account for the next algorithm. I assume that is a change we can safely make later?

Reply:
I think in this particular case, the toil of potentially changing this in the future if for some reason a change to the OCI spec comes along, is low enough risk and minor enough that I'd encourage the validation now, the validation benefits outweigh the risks and toil to me

We can never be perfect about not requiring "breaking" changes in the future, but in general, relaxing validation is easier for us since we generally own the consumers